### ISSUE TITLE
Custom action fails because Fork starts another git instance when custom action is not yet finished

### ISSUE THREAD

**Issue by lvernaillen** [ForkIssues/TrackerWin#512]
Hi,

I have a custom action that performs some git commands (git rebase among others).
Our workflow requires to "copy-rebase" a feature branch to another location while preserving the original feature branch.
This custom action regularly fails with **"Another git process seems to be running in this repository"**.

This is because Fork starts another git process (git status or git log) eventhough the custom action has not finished yet.
I would expect Fork to wait for the custom action to exit before Fork starts it's own git actions.
The "Wait for exit" option is checked in the custom action settings.

Tried to disable "Fetch remotes automatically" but that doesn't make a difference.
If I run the script of that custom action outside of Fork (when Fork is not running), this does not occur.

Might be related to #181 

This is seen in Fork 1.40.2.0 on Windows 10.

Thanks in advance for looking into it!
Lieven

**Comment by DanPristupov** [comment]
> This custom action regularly fails with "Another git process seems to be running in this repository".

I suppose it means that the repository is **already** busy, otherwise your 'copy-rebase' would lock it first.

This might be related to the `git status` command.
1. Try to disable 'tab indicator for uncommitted changes' in preferences.
2. Ensure that the busy indicator in the status control is not spinning when you start the operation
3. run `git gc` to improve the `git status` performance.

**Comment by lvernaillen** [comment]
Thanks for your quick reply.

Disabling _'tab indicator for uncommitted changes'_ in preferences indeed prevents some occurences of those errors, while missing out on the automatic status updates.

However, Fork also seems to run a `git status` when it's window gets focus again, even with that preference disabled.
So when focus is on one window (example my IDE on left screen), then I move my cursor over Fork's window (on right screen) and trigger the custom action, I get that error again.

I know, this is your point 2:
> Ensure that the busy indicator in the status control is not spinning when you start the operation.

 but it feels strange to switch to Fork's window and have to wait few seconds before doing an action.

The custom action performs several git actions. First creating a temp branch so the original "to copy" branch isn't changed. Then rebasing that temp branch, squashing commits, ....
I don't know if the failure is that Fork's `git status` tries to run while the repo is already busy because of a git command inside the custom action. Or it is the other way around that the custom action cannot perform one of it's git commands because Fork's `git status` is running just in between some git actions of custom command.

- Is it possible to disable the `git status` action when Fork gets focus in case the  _'tab indicator for uncommitted changes'_ is disabled?

- Or maybe better, but probably harder to implement, prevent Fork performing git actions when a custom action with _'wait for exit'_ is running? This would allow to leave _'Fetch remotes automatically'_ and _'tab indicator for uncommitted changes'_ enabled.

**Comment by lvernaillen** [comment]
Hi,

The custom actions really aren't usable for scripting.
Fork starts a lot of Git instances in parallel resulting in conflicts.
So even with your suggestions it's a hit-and-miss and we can't use them from the Fork app.

To clarify the issue, these are the git instances Fork starts while the custom action is still running:
![fork_git_instances](https://user-images.githubusercontent.com/4180374/69135232-11ad0280-0ab9-11ea-99fe-5ac569990f01.jpg)

If Fork app is running but I start that same script from the command line, instead of as a custom action in Fork, then **no** Fork git instances are started and the script succeeds without any git concurrency issues.

I can see the difference between Fork's git instances and my script's git instances because my script uses git installed on the system, while Fork uses git that is bundled with Fork and resides in AppData\Local\Fork\gitInstance.

Expected behavior:
Fork shouldn't start git actions when a custom action configured with "wait for exit" is running.

What is your opinion on this?
Thanks in advance.

**Comment by DanPristupov** [comment]
> Fork **shouldn't start git actions** when a custom action configured with "wait for exit" is running.

Do you mean Fork should block the UI? You can do that yourself by enabling 'show output'.

**Comment by lvernaillen** [comment]
No, I mean that Fork shouldn't start git instances behind the scenes as long as the custom command hasn't finished. That is why "Wait For Exit" is checked. As soon as the script exists, Fork can do whatever git actions to update the UI to the current git state.

"Wait For Exit" and "Show Output" are already both checked in my scenario.
"Fetch remotes automatically" and "tab indicator for uncommitted changes" are unchecked in the preferences.

But apparently Fork starts several git actions while the custom command is running. See screenshot in previous post. These Fork initiated git actions then introduce race condition. So it seems as if Fork is monitoring the repo behind the scenes and starts some actions because of changes in the repo.

However if I start that same script from the command line, outside of Fork but with Fork app still running, then it's fine. Same changes occur to the repo, but Fork doesn't start it's own git actions. This same behavior is what I expect when I start the script using a custom command.

So for some reason if the script is started from a custom action, Fork is triggered and starts it's own concurrent git actions. This shouldn't happen. Fork should wait until the script exists.

To monitor the git actions started by the script or by Fork I used [Procmon ](https://docs.microsoft.com/en-us/sysinternals/downloads/procmon) tool from SysInternals. Filter on
- "Process Name" contains "git"
- "Operation" is "Process Start"

I used the git path to distinguish between script git instances and Fork git instances.
You could also use the Parent PID.

**Comment by DanPristupov** [comment]
>"Wait For Exit" and "Show Output" are already both checked in my scenario.

Yes, you are right, sorry. I forgot, the output appears when the operation is done.

> But apparently Fork starts several git actions while the custom command is running.

No, Fork only starts refresh on focus or after a change. Otherwise it's a bug.

> Fork shouldn't start git actions when a custom action configured with "wait for exit" is running.

How can I reproduce the problem? Can you create a small script please? May be the solution could be less harsh. There are a lot of people who use custom actions for different cases and don't need to stop UI updates.

**Comment by lvernaillen** [comment]
I created a Powershell script that does a "copy-rebase and squash" of a selected feature branch to your local head. Can be used to copy a feature from feature branch to currently active qa branch and squash it to one commit containing the author info. There is also a .bat file provided to prevent Powershell permission issues. It should have the same name as the Powershell script.

[CustomActionTest.zip](https://github.com/ForkIssues/TrackerWin/files/3864128/CustomActionTest.zip)

Reproduce on Windows (I use Win10):

- Place the .ps1 and .bat file in a folder on your pc
- In Fork create a custom action on a branch
- Choose the .bat file as script path
- As parameters use: `-featBranch $name`
- Check "Wait For Exit" and "Show Output"
- Create a feature branch based on master with multiple commits
- Create a qa branch based on master with other commits (prevent merge conflicts, this concise version is not compatible with merge conflicts)
- Have qa branch checked out
- Have Procmon running with the filter specified above to see Fork's git actions
- Right mouse click on the feature branch
- Run the custom command

If all is well, you have your feature branch commits squashed in one commit on top of current head.
This will sometimes work and sometimes not. On my (faster) pc I have more luck. Colleague of mine has slower pc and has more errors.

Note: There is a form that pops up in this script. It is used to ask user for confirmation of the action. In this demo script it has no text only an OK and Cancel button. What I noticed making this example is that Fork app starts git actions when the popup form is closed (after you click OK)! If you repeat the `$result = $form.ShowDialog()` line, you get multiple popups sequentially. Each time Fork starts git actions.

I seems that these Fork actions interfere with the git actions in the script.
This might be source of the bug!
Why does Fork start git actions on each close of a popup, while script isn't finished yet?
Do you recognize these git actions? Why are they triggered before end of the custom action?

Maybe these Fork git actions are also triggered after the editor to choose a commit message for the squashed commit closes.

>  Fork only starts refresh on focus or after a change.

What do you mean with "after a change"? After what change?

Looking forward to your insights.

**Comment by lvernaillen** [comment]
Sorry, I clicked "Close and comment" instead of "Comment".

**Comment by DanPristupov** [comment]
>  After what change?

I meant an operation made by user interaction. Checkout, stage, button click, etc.

So, I'm running your script now. I couldn't reproduce the problem yet, however I see that the refresh is caused by the prompt dialog in the script.

As I already mentioned:
> Fork only starts refresh on focus

So, don't show the dialog to not fire the OnActivated event in Fork when it's closed and Fork will not make the refresh.

**Comment by lvernaillen** [comment]
That's a pitty.
This basically means that Fork doesn't support interactive custom commands...

Even with my form dialog popups removed, this issue occurs. Since some git actions trigger a texteditor to open. So those git actions aren't supported either in custom commands.

See this modified script as an example:
[ForkCustomActionIssueNoForms.zip](https://github.com/ForkIssues/TrackerWin/files/3875811/ForkCustomActionIssueNoForms.zip)

(EDIT: use the example in my next comment, it's a very simple one causing the same error)

Because the script performs a merge squash and commits those, the git configured editor opens automatically to specify a commit message for that squash commit. As soon as that editor is closed, Fork is triggered to perform git refresh actions. Those interfere with the script resulting in an error.

Reproduce on Windows (I use Win10):
- Place the .ps1 and .bat file in a folder on your pc
- In Fork create a custom action on a branch
- Choose the .bat file as script path
- As parameters use: -featBranch $name
- Check "Wait For Exit" and "Show Output"
- Create a feature branch based on master with *multiple* commits
- Create a qa branch based on master with other commits (prevent merge conflicts, this concise - version is not compatible with merge conflicts)
- Have qa branch checked out
- Right mouse click on the feature branch
- Run the custom command
- Watch it fail

Personally I'd prefer Fork to really wait for the custom action to finish before triggering any git actions. That would open up more custom action possibilities.

**Comment by lvernaillen** [comment]
To trigger this issue the custom action script can even be very simple.
Trying to merge squash a feature branch to HEAD and then adding some commits.

[ForkCustomActionIssueMergeSquash.zip](https://github.com/ForkIssues/TrackerWin/files/3875892/ForkCustomActionIssueMergeSquash.zip)

This example makes it much more clear.
- Create a first branch based with multiple commits
- Create a second branch with other commits (prevent merge conflicts with first branch)
- Have second branch checked out
- Right mouse click on the first branch
- Run the custom command to merge squash this to HEAD
- Watch it fail
